core/code/ornaments.js

```diff
@@ -13,52 +13,120 @@
     }
-    this._layer = layerGroup();
-    this._beacons = layerGroup();
-    this._frackers = layerGroup();
+    this.load();
+
+    this._layer = window.ornaments.layerGroup();
+    this._excluded = window.ornaments.layerGroup(); // to keep excluded ornaments in an own layer
+
     window.layerChooser.addOverlay(this._layer, 'Ornaments');
-    window.layerChooser.addOverlay(this._beacons, 'Beacons');
-    window.layerChooser.addOverlay(this._frackers, 'Frackers');
+    window.layerChooser.addOverlay(this._excluded, 'Excluded', {default: false});
+
+    $('<a>')
+      .html('Ornaments Opt')
+      .attr({
+        id: 'ornaments-toolbox-link',
+        title: 'Edit ornament exclusions',
+        accesskey: 'o'
+      })
+      .click(window.ornaments.ornamentsOpt)
+      .appendTo('#toolbox');
+  },
+  createLayer: function (layerID) {
+    window.ornaments[layerID] = window.ornaments.layerGroup();
+    window.layerChooser.addOverlay(window.ornaments[layerID], layerID);
   },
 
   addPortal: function (portal) {
     this.removePortal(portal);
-
     var ornaments = portal.options.data.ornaments;
     if (ornaments && ornaments.length) {
       this._portals[portal.options.guid] = ornaments.map(function (ornament) {
         var layer = this._layer;
-        if (ornament.startsWith('pe')) {
-          layer = ornament === 'peFRACK'
-            ? this._frackers
-            : this._beacons;
-        }
+        var opacity = this.OVERLAY_OPACITY;
         var size = this.OVERLAY_SIZE;
+        var anchor = [size / 2, size / 2];
+        var iconUrl = '//commondatastorage.googleapis.com/ingress.com/img/map_icons/marker_images/' + ornament + '.png';
+
+        if (!this.knownOrnaments[ornament]) {
+          this.knownOrnaments[ornament]=false;
+        }
+
+        if (typeof (window.ornaments.icon[ornament]) !== 'undefined') {
+          if (window.ornaments.icon[ornament].layer) {
+            if (window.ornaments[window.ornaments.icon[ornament].layer] === undefined){
+              console.log ('Add missing layer: ',window.ornaments.icon[ornament].layer);
+              window.ornaments.createLayer(window.ornaments.icon[ornament].layer);
+            }
+            layer =  window.ornaments[window.ornaments.icon[ornament].layer];
+          }
+          opacity = 1;
+          if (window.ornaments.icon[ornament].url) {
+            iconUrl = window.ornaments.icon[ornament].url;
+            if (window.ornaments.icon[ornament].offset) {
+              switch (window.ornaments.icon[ornament].offset) {
+              case 1:
+                anchor = [size / 2, size];
+                break;
+              case 0:
+                anchor = [size / 2, size / 2];
+                break;
+              case -1:
+                anchor = [size / 2, - size ];
+              }
+            }
+          }
+        }
+
+        var exclude = false;
+        if (this.excludedOrnaments && this.excludedOrnaments !== ['']) {
```

Comment:
wouldn't it be easier to make sure "excludedOrnaments" is in a valid state?

Reply:
without "Safety checks" code is always easier. 
The content depends on the users's input.
An empty string is a valid input and results in `['']`.
I think the rest of the code would be more complex.

Reply:
`this.excludedOrnaments !== ['']` is _always_ true, even if `this.excludedOrnaments` is an array with a single element `''`
the input already filter out empty strings, so you just need to check whether the array is empty

```suggestion
        if (this.excludedOrnaments.length) {
```